Allow docker image version to be customized

tlvenn opened the pull request:
Closes #4

tlvenn committed:
Allow docker image version to be customized

Closes #4

teamon commented on lib/mix_docker.ex:
```diff
@@ -80,15 +88,19 @@ defmodule MixDocker do
     sha     = git_head_sha()
 
-    "#{version}.#{count}-#{sha}"
+    Code.eval_string(version_template, mix_version: version, git_count: count, git_sha: sha) |> elem(0)
```
It should be simple replace "${var}" with value, no need to use dangerous eval_string here.

tlvenn replied:
Ya I wasn't sure exactly how to best handle this. So I guess I will change it to use a regexp and match on `${var}`

tlvenn replied:
Done

teamon commented on lib/mix_docker.ex:
```diff
@@ -80,15 +88,19 @@ defmodule MixDocker do
+  defp image_version(args) do
+    version = (OptionParser.parse(args) |> elem(0) |> Keyword.get(:version, Application.get_env(:mix_docker, :version, "\#{mix_version}.\#{git_count}-\#{git_sha}")))
+    "\"" <> version <> "\""
+  end
```
This is a bit hard to read, please split it into multiple lines. 
It should be obvious that CLI option has precedence over configuration (and is both are missing  we use the default)

tlvenn replied:
Will do, going with the regexp approach will simplify this as well as there is no need to escape the string anymore.

tlvenn replied:
Done

teamon reviewed (changes requested):
Thanks a lot, this is great! 
Please see my comments below.

(NOTE: I'm leaving for holidays until 20.03 and I won't be able to respond)

tlvenn commented:
Will fix the above very soon, enjoy your holidays @teamon  !

tlvenn committed:
Use simple String replacements instead of Code.eval_string

tlvenn committed:
Properly test if the project is in git already

tlvenn committed:
Filter the args passed down to docker build.

Given the args are passed down to docker build and that Docker emits an error on a arg it does not know, we need to filter out our own args to keep Docker happy.

tlvenn commented:
Hi @teamon, this should be ready to merge now.

tlvenn commented:
Hi @teamon , hope you had good holidays. Any chance you can take a look at my changes and my other PR ? Thanks in advance.

tlvenn commented:
Hi @teamon, I know you must be busy but a sign of life would be greatly appreciated. Thanks in advance.

teamon commented:
Hey @tlvenn, thanks for patience!

I merged your PR and then changed it a bit, because:
- The `"${var}"` syntax was interpolated by shell - I ditched the `$`, so now it's just `{var}`
- I replaces OptionParser with simple recursive extraction - this allowed to simplify the flow of arguments
- Instead of checking for .git directory the git commands are now only executed if the template references them - this way the proper error will be raised (instead of silent empty string)

The versions 0.4.0 has been released. Please do let me know if you have any feedback.
Thanks.
